File: backends/fastapi/app/api/v1/me.py
"""GET/PATCH /me · POST /me/password · GET /me/capsules · DELETE /me/capsules/{id}"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query, Response
from sqlalchemy.orm import Session

from app.deps import current_user_required, get_db
from app.models import User
from app.schemas.capsule import CapsuleListItem
from app.schemas.common import Envelope, Paginated
from app.schemas.user import ChangePasswordRequest, UpdateProfileRequest, UserOut
from app.services import auth_service, capsule_service, plaza_service, user_service

router = APIRouter(prefix="/me", tags=["Me"])


@router.get("", response_model=Envelope[UserOut])
def get_me(user: User = Depends(current_user_required)) -> Envelope[UserOut]:
    return Envelope(success=True, data=user_service.to_out(user))


@router.patch("", response_model=Envelope[UserOut])
def patch_me(
    req: UpdateProfileRequest,
    db: Session = Depends(get_db),
    user: User = Depends(current_user_required),
) -> Envelope[UserOut]:
    out = user_service.update_profile(db, user=user, req=req)
    return Envelope(success=True, data=out)


@router.post("/password", status_code=204, response_class=Response)
def post_password(
    req: ChangePasswordRequest,
    db: Session = Depends(get_db),
    user: User = Depends(current_user_required),
) -> Response:
    auth_service.change_password(
        db, user=user, current_password=req.currentPassword, new_password=req.newPassword
    )
    return Response(status_code=204)


@router.get("/capsules", response_model=Envelope[Paginated[CapsuleListItem]])
def get_my_capsules(
    page: int = Query(default=1, ge=1),
    pageSize: int = Query(default=20, ge=1, le=50),
    db: Session = Depends(get_db),
    user: User = Depends(current_user_required),
) -> Envelope[Paginated[CapsuleListItem]]:
    data = plaza_service.my_capsules(db, user=user, page=page, page_size=pageSize)
    return Envelope(success=True, data=data)


@router.delete("/capsules/{capsule_id}", status_code=204, response_class=Response)
def delete_my_capsule(
    capsule_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(current_user_required),
) -> Response:
    capsule_service.delete_own(db, user=user, capsule_id=capsule_id)
    return Response(status_code=204)

File: backends/gin/internal/handler/me.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"hellotime/gin/internal/core"
	"hellotime/gin/internal/dto"
	"hellotime/gin/internal/middleware"
	"hellotime/gin/internal/service"
)

// GetMe GET /api/v1/me
func GetMe(c *gin.Context) {
	user := middleware.CurrentUser(c)
	c.JSON(http.StatusOK, dto.OK(service.GetMe(user)))
}

// PatchMe PATCH /api/v1/me
func PatchMe(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		user := middleware.CurrentUser(c)
		var req dto.UpdateProfileRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			middleware.RespondErr(c, core.ValidationErr(err.Error(), "body"))
			return
		}
		if req.Nickname == nil && req.AvatarID == nil {
			middleware.RespondErr(c, core.ValidationErr("nickname / avatarId 至少提供一个", "body"))
			return
		}
		out, err := service.UpdateProfile(db, user, req)
		if err != nil {
			middleware.RespondErr(c, err)
			return
		}
		c.JSON(http.StatusOK, dto.OK(out))
	}
}

// PostPassword POST /api/v1/me/password
func PostPassword(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		user := middleware.CurrentUser(c)
		var req dto.ChangePasswordRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			middleware.RespondErr(c, core.ValidationErr(err.Error(), "body"))
			return
		}
		if err := service.ChangePassword(db, user, req.CurrentPassword, req.NewPassword); err != nil {
			middleware.RespondErr(c, err)
			return
		}
		c.Status(http.StatusNoContent)
	}
}

// GetMyCapsules GET /api/v1/me/capsules
func GetMyCapsules(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		user := middleware.CurrentUser(c)
		page, err := queryInt(c, "page", 1)
		if err != nil {
			middleware.RespondErr(c, err)
			return
		}
		pageSize, err := queryInt(c, "pageSize", 20)
		if err != nil {
			middleware.RespondErr(c, err)
			return
		}
		data, err := service.MyCapsules(db, user, page, pageSize)
		if err != nil {
			middleware.RespondErr(c, err)
			return
		}
		c.JSON(http.StatusOK, dto.OK(data))
	}
}

// DeleteMyCapsule DELETE /api/v1/me/capsules/:id
func DeleteMyCapsule(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		user := middleware.CurrentUser(c)
		capsuleID := c.Param("id")
		if err := service.DeleteOwnCapsule(db, user, capsuleID); err != nil {
			middleware.RespondErr(c, err)
			return
		}
		c.Status(http.StatusNoContent)
	}
}
